### PR TITLE
Address issue #291 by adding the formatter control

### PR DESCRIPTION
**Pull request by gltrost**
The following branch creates a new flag called `formatter` that allows the user to speficy if the results should be printed to `stderr` or `stdout`. If the flag is not set, the results default to `stderr`. 

This requires changes to `Output.print_result` and `Precondition.check`, which both print information. The function `Precondition.exclude`  was also altered to take into account the changes to `check` since `exclude` calls `check`.

### CODE REVIEW

**Review by jtpaasch — CHANGES_REQUESTED** [review]
We don't need to change the CLI plugin. Giving the CLI user the option to print to stdout or stderr doesn't give them anything they don't already have with standard redirection techniques. 

So I would undo any of the changes in the `cbat_tools/wp/plugin` folder (e.g., undo the changes to `wp.ml`, and to parameters, and take out the bit that validates the `--formatter` param, etc.)

All we need is to give users of the _library_ the ability to specify where they want the output to go.

In the library functions, we should be able to let the user pass in a formatter directly, via a parameter `?fmt:(fmt = Format.stderr_formatter)`, and then we should be able to use `Format.fprintf fmt` to print things. E.g., I would alter `print_result` to be like this:

```
let print_result (solver : Solver.solver) (status : Solver.status) (goals: Constr.t)
    ?fmt:(fmt = Format.err_formatter)
    ~show:(show : string list) ~orig:(env1, sub1 : Env.t * Sub.t)
    ~modif:(env2, sub2 : Env.t * Sub.t) : unit =
  | Solver.UNSATISFIABLE -> Format.fprintf fmt "%s%!" "\nUNSAT!\n"
  | Solver.UNKNOWN -> Format.fprintf fmt "%s%!" "\nUNKNOWN!\n"
  | Solver.SATISFIABLE -> ...
```

And `check` would go like this:

```
let check ?fmt:(fmt = Format.err_formatter) ?refute:(refute = true) ?(print_constr = []) ?(debug = false)
    (solver : Solver.solver) (ctx : Z3.context) (pre : Constr.t)  : Solver.status =
  Format.fprintf fmt "Evaluating precondition.\n%!";
  if (List.mem print_constr "precond-internal" ~equal:(String.equal)) then (
    Format.fprintf fmt "Internal : %s \n %!" (Constr.to_string pre) ) ;
  let pre' = Constr.eval ~debug:debug pre ctx in
  Format.fprintf fmt "Checking precondition with Z3.\n%!";
  ... 
```

And so on. Wouldn't that work?

**Comment by gltrost** [comment]
> And so on. Wouldn't that work?
@jtpaasch Yeah! That or something very close to that, should work. I'll make those changes now.

File: wp/lib/bap_wp/src/output.mli
Comment (jtpaasch) [inline wp/lib/bap_wp/src/output.mli:35]: Can we write `Format.formatter` instead of `Stdlib__format.formatter`? I don't have any opinion on whether we should or should not do that. I'm just curious if there's a reason why you fully specified `Stdlib_format` instead of just `Format`. And by no means should this hold up merging this.

**Review by jtpaasch — APPROVED** [review]
Everything looks good to me. I made one comment, mostly out of curiosity, which you're welcome to look at or ignore. After the merge conflicts are fixed, I say merge it.

File: wp/lib/bap_wp/src/output.mli
Comment (gltrost) [reply wp/lib/bap_wp/src/output.mli:35]: There's no deeper reason for using `Stdlib__format` instead of `Format`, and using `Format` would be more consistent. I'll change this now.

**Pull request merged**
